src/PreconditionerTypes.f90

```diff
@@ -343,6 +550,695 @@ SUBROUTINE setup_ILU_PreCondtype(thisPC)
     ENDIF
   ENDIF
 ENDSUBROUTINE setup_ILU_PreCondtype
+!
+!-------------------------------------------------------------------------------
+!> @brief Intializes up the SOR Preconditioner Type with a parameter list
+!> @param thisPC The preconditioner to act on
+!> @param A The matrix to precondition
+!> @param params The parameter list
+!>
+SUBROUTINE init_SOR_PreCondtype(thisPC,A,params)
+  CHARACTER(LEN=*),PARAMETER :: myName='init_RSOR_PreCondType'
+  CLASS(SOR_PrecondType),INTENT(INOUT) :: thisPC
+  CLASS(MatrixType),ALLOCATABLE,TARGET,INTENT(IN),OPTIONAL :: A
+  TYPE(ParamType),INTENT(IN),OPTIONAL :: params
+  TYPE(ParamType)::PListMat_LU
+  INTEGER(SIK)::k
+
+  !required statuses
+  REQUIRE(.NOT. thisPC%isinit)
+  REQUIRE(PRESENT(A))
+  REQUIRE(ALLOCATED(A))
+  REQUIRE(A%isInit)
```

Comment:
Why require this if `A` is declared `OPTIONAL`?

Reply:
The abstract interface specifies A as optional, so this subroutine must match. However, the SOR preconditioners require A to be present, so we have to require that A is passed in.

Reply:
Can we make it required in the interface?  Are there preconditioners that don't require a matrix?

Reply:
I did a brief glance and it seems that all of them assume the presence of A.  And quite frankly, I think it's very reasonable to assume that when setting up a preconditioner, a matrix is also available.  Dan, I think your suggestion is right.  Let's just make it required.  If some preconditioner happens to ignore it (I don't even know how that would be possible, really), then that's fine.

Reply:
Quoting from Matt:
Regarding the optional matrix question, the answer is that unfortunately one of the preconditioner types implemented in Futility is a Shell preconditioner for PETSc, which does not have a Futility matrix but a PETSc one. Since these are not type-compatible, we cannot make the matrix a requirement.

Reply:
@mattkab2 I see 'PETSC_PCSHELL_PC` in `PreconditionerTypes` which is a `CLASS(PreConditionerType)`, which owns a `CLASS(MatrixType),POINTER :: A`.  There is also a `PETScMatrixType` which has the futility `MatrixType` as its base class.  I'm not sure I'm understanding what the issue is?

Reply:
I've dug a little deeper, because your comment is correct. The trouble actually comes from the intended use case of the SHELL_PC. From what I gather in the unit tests, the intent is to allow the user to define the intended init/clear/setup/apply functions. This may be confusing or strange but it is the functionality expected and tested by the unit tests.

Within the unit test, the preconditioner is initialized without a matrix, which is set only after the fact (see lines 440-560 in the testPreconditioner unit test). Since this is the one and only test of PCShell I have to assume that this is the intended use case and the preconditioner must be able to be initialized without an A.

If we want to change this, I'm open to it but I'm not sure what repercussions it might have to other parts of the code. They are probably minimal, but I'm not sure. I'm also not certain it should be part of this work.

Reply:
I did some searching.  I expect virtually 0 side effects.  However, you're right that it doesn't need to go in with this work.  I may do a follow-up merge to clean up the weird interface.